nep-Y.mediawiki

```diff
@@ -0,0 +1,324 @@
+specification (from -32700 to -32600) for any calls when they're
+appropriate. The most often used one is -32602 "Invalid params", it SHOULD be
+used only when call parameters are syntactically incorrect and servers
+SHOULD give additional data like offending parameter number and why it's wrong
+in the <code>data</code> response field.
+
+====Groups====
+
+Error codes are grouped. Each call can use any appropriate groups and codes as
+long as they fit the purpose.
+
+{|
+|+ Error code ranges
+|-
+! Range           !! Usage
+|-
+| -15000...-10000 || Experimental implementation-defined range.
+|-
+| -699...-600     || Configuration and service errors.
+|-
+| -599...-500     || Inventory verification errors.
+|-
+| -399...-300     || Wallet-related errors.
+|-
+| -199...-100     || Missing/unknown items/inventories/states.
+|}
+
+Experimental codes can be used by node implementations if they're documented,
+however clients can't rely on the meaning of these codes if the version of the
+node changes or if they connect to a different node implementation. New
+proposals can set new ranges or specific codes as long as they don't conflict
+with this specification. All the other codes are reserved for future use and
+MUST NOT be used by compliant implementations.
+
+====Missing items====
+
+{|
+|+ Codes for missing items
+|-
+! Code !! Message            !! Meaning
+|-
+| -101 || Unknown block      || Call that accepts as a parameter or searches for a header or a block as a part of its job can't find it.
+|-
+| -102 || Unknown contract   || Call that accepts as a parameter or searches for a contract as a part of its job can't find it.
+|-
+| -103 || Unknown transaction|| Call that accepts as a parameter or searches for a transaction as a part of its job can't find it.
+|-
+| -104 || Unknown storage item|| Call that looks for an item in the storage can't find it.
+|-
+| -105 || Unknown script containter|| Call that accepts as a parameter or searches for a script container (a block or transaction) as a part of its job can't find it (this error generalizes -101 and -103 in cases where it's needed).
+|-
+| -106 || Unknown state root || Call that accepts as a parameter or searches for a state root as a part of its job can't find it.
+|-
+| -107 || Unknown session    || Call that accepts as a parameter or searches for a session as a part of its job can't find it.
+|-
+| -108 || Unknown iterator   || Call that accepts as a parameter or searches for a session iterator as a part of its job can't find it.
+|-
+| -109 || Unknown height     || Block or header height passed as parameter or calculated during call execution is not correct (out of the range known to the node).
+|}
+
+
+====Wallet-related problems====
+
+{|
+|+ Codes for calls that use a wallet
+|-
+! Code !! Message            !! Meaning
+|-
+| -300 || Insufficient funds || Transaction that sends some assets can't be created because it fails.
```

Comment:
The problem at the moment is that this code is returned from `send*` calls exactly for this reason: failure to create a transaction. In 99% cases that's because the account doesn't have enough balance, but at the same time it can also happen when wrong signers are used or when `onNEP17Payment` fails or something else. So either implementation needs to be changed to separate "insufficient funds" from generic "execution error" (-608, btw) or we have this vague error code. I'm not against using both -300 and -608 for `send*`, but I'm not sure what would be the consensus about it.

Reply:
I hope that with this specification the implementations are updated to take away this ambiguity in their error responses. So I'd be in favour of separating, but as you said there needs to be consensus among the node implementations. 

If no consensus is reached on separating, then I think we should at least not name it "Insufficient funds" but something more generic and then use the `data` field to clarify if it is a balance issue, wronger signers etc.

Reply:
It's easy to support this separation from NeoGo side, we don't (and won't) have `send*` implementations.

Reply:
Is there any other opinions on this topic? It’s OK to me to have this generic -300 code, because we always may specify some additional data.